src/orderbook/index.ts

```diff
@@ -0,0 +1,258 @@
+/**
+ * Module containing event based orderbook reading implementation. The entire
+ * orderbook is contructed from Ethreum events fetched in two stages: first,
+ * the past events are fetched in block-based pages, second an event subsciption
+ * is setup to listen to new incomming events that get added to the orderbook
+ * snapshot.
+ *
+ * The streamed orderbook can be queried at any time to get the current open
+ * orders (that is orders and balances as it would they would be included for
+ * the next batch) or the finalized orderbook (orders and balances that are
+ * being considered for trading by the solver).
+ *
+ * @packageDocumentation
+ */
+
+import assert from 'assert'
+import Web3 from 'web3'
+import { TransactionReceipt } from 'web3-core'
+import { EventData } from 'web3-eth-contract'
+import * as BatchExchangeArtifact from '@gnosis.pm/dex-contracts/build/contracts/BatchExchange.json'
+import { BatchExchange } from '@gnosis.pm/dex-contracts/build/types/BatchExchange'
+import { AccountState } from './state'
+
+/**
+ * Configuration options for the streamed orderbook.
+ */
+export interface OrderbookOptions {
+  /**
+   * Optinally specify the last block to apply events for. This is useful for
+   * testing the streamed orderbook and ensuring that it is producing a correct
+   * account state for a known block.
+   */
+  endBlock?: number;
+
+  /**
+   * Set the block page size used to query past events.
+   *
+   * @remarks
+   * Nodes are usually configured to limit the number of events that can be
+   * returned by a single call to retrieve past logs (10,000 for Infura)
+   * so this parameter should be set accordingly.
+   */
+  blockPageSize: number;
+
+  /**
+   * Enable strict checking that performs additional integrety checks.
+   *
+   * @remarks
+   * The additional integrety checks have a non-negligeable runtime cost so
+   * they are disabled by default, but can help diagnose issues and bugs in the
+   * streamed orderbook.
+   */
+  strict: boolean;
+}
+
+/**
+ * The default orderbook options.
+ */
+export const DEFAULT_ORDERBOOK_OPTIONS: OrderbookOptions = {
+  blockPageSize: 10000,
+  strict: false,
+}
+
+/**
+ * The duration in seconds of a batch.
+ */
+export const BATCH_DURATION = 300
+
+/**
+ * A cached retriver of batch IDs for specific blocks and events.
+ */
+export class BatchIds {
+  private readonly cache: Map<number, number> = new Map();
+
+  constructor(
+    private readonly web3: Web3,
+  ) {}
```

Comment:
I see from above that having useless constructors is disabled. Is this considered a useless constructor? What information does it convey?

Reply:
Its because typescript allows constructors that automatically set properties. So:
```
private readonly web3: Web3
constructor(web3: Web3) {
  this.web3 = web3
}
```
is equivalent to:
```
constructor(
  private readonly web3: Web3
) {}
```

So the linter gets confused and thinks this constructor does nothing, but it does!